File: ios/StableChannels/StableChannels/Services/WebSocket/ReconnectionManager.swift
import Foundation
import os.log

/// Manages WebSocket reconnection with exponential backoff and a keepalive ping timer.
@MainActor
final class ReconnectionManager: NSObject {
    var reconnectAttempts: Int = 0
    let maxReconnectDelay: UInt64 // seconds
    private var reconnectTask: Task<Void, Never>?
    private var pingTimer: Timer?
    var isManualDisconnect: Bool = false

    /// Called every 30s — the service should send a WebSocket ping frame.
    var onPingRequest: (() -> Void)?

    /// Called when the backoff delay expires — the service should call `connect()`.
    var onReconnect: (() -> Void)?

    private let logger = Logger(subsystem: "com.stablechannels", category: "reconnect")

    init(maxReconnectDelay: UInt64 = 60) {
        self.maxReconnectDelay = maxReconnectDelay
    }

    // MARK: - Lifecycle

    /// Connection is alive — reset attempts and start the ping timer.
    func connected() {
        reconnectAttempts = 0
        stopReconnectTask()
        startPingTimer()
    }

    /// Connection dropped unexpectedly — stop ping and schedule reconnect.
    func disconnected() {
        stopPingTimer()
        if isManualDisconnect { return }
        scheduleReconnect()
    }

    /// Graceful shutdown — stop everything.
    func stop() {
        isManualDisconnect = true
        stopReconnectTask()
        stopPingTimer()
    }

    /// Reset manual-disconnect flag before a fresh `connect()`.
    func reset() {
        isManualDisconnect = false
        reconnectAttempts = 0
    }

    // MARK: - Ping Timer

    private func startPingTimer() {
        stopPingTimer()
        pingTimer = Timer.scheduledTimer(withTimeInterval: 30.0, repeats: true) { [weak self] _ in
            guard let self else { return }
            self.onPingRequest?()
        }
    }

    private func stopPingTimer() {
        pingTimer?.invalidate()
        pingTimer = nil
    }

    // MARK: - Reconnection Backoff

    private func scheduleReconnect() {
        stopReconnectTask()

        let delay = min(UInt64(pow(2.0, Double(reconnectAttempts))), maxReconnectDelay)
        reconnectAttempts += 1

        logger.info("Scheduling reconnect in \(delay)s (attempt \(self.reconnectAttempts))")

        reconnectTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: delay * 1_000_000_000)
                await MainActor.run {
                    guard let self, !self.isManualDisconnect else { return }
                    self.onReconnect?()
                }
            } catch {
                // Task was cancelled
            }
        }
    }

    func stopReconnectTask() {
        reconnectTask?.cancel()
        reconnectTask = nil
    }
}

File: android/app/src/main/java/com/stablechannels/app/services/websocket/ReconnectionManager.kt
package com.stablechannels.app.services.websocket

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.min
import kotlin.math.pow

class ReconnectionManager(
    private val scope: CoroutineScope,
    private val maxReconnectDelaySeconds: Long = 60L,
    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
) {
    @Volatile
    var reconnectAttempts: Int = 0
        private set

    @Volatile
    var isManualDisconnect: Boolean = false

    private var reconnectJob: Job? = null

    var onReconnect: (() -> Unit)? = null

    fun connected() {
        reconnectAttempts = 0
        stopReconnectTask()
    }

    fun disconnected() {
        if (isManualDisconnect) {
            return
        }
        scheduleReconnect()
    }

    fun stop() {
        isManualDisconnect = true
        stopReconnectTask()
    }

    fun reset() {
        isManualDisconnect = false
        reconnectAttempts = 0
    }

    private fun scheduleReconnect() {
        stopReconnectTask()

        val delaySeconds = min((2.0.pow(reconnectAttempts.toDouble())).toLong(), maxReconnectDelaySeconds)
        reconnectAttempts += 1

        reconnectJob = scope.launch(dispatcher) {
            delay(delaySeconds * 1000)
            if (!isManualDisconnect) {
                onReconnect?.invoke()
            }
        }
    }

    fun stopReconnectTask() {
        reconnectJob?.cancel()
        reconnectJob = null
    }
}
